midend/replaceSelectRange.cpp

```diff
@@ -115,12 +143,47 @@ DoReplaceSelectRange::cartesianAppend(const std::vector<IR::Vector<IR::Expressio
     return newVecs;
 }
 
+const IR::Node *DoReplaceSelectRange::preorder(IR::SelectExpression *e) {
+    BUG_CHECK(!inSelect, "A select nested in select: %1%", e);
+    inSelect = true;
```

Comment:
In general you don't need such a boolean flag.
It can be replaced with `findContext<IR::SelectExpression>() != nullptr`, called from the child node.

Reply:
This flag is there mostly to avoid introduction of more inconsistencies if AST is unexpected (or visitor is broken, which is probably significantly less likely). We could avoid it altogether but I believe a cheap assert is better then possibility on long debugging in future. If I replace it with `findContext` I don't think it would be that cheap.

Reply:
The compiler code is not optimized for speed, but rather for correctness, readability, and maintainability. I think that findContext is more readable, and it is in only one place, so it is more maintainable.

Reply:
changed